### DISCUSSION TITLE
Recording Evidence and Status of a Satisfied Control?

### DISCUSSION THREAD

**Discussion by flickerfly** [usnistgov/OSCAL#1074] [Q&A] [answered]
I'm trying to wrap my head around the Assessment Results model. I'm assuming this is where I'd put information about how I've tested for and confirmed compliance to a specific control, but I'm struggling to wrap my head around it.

Would this type of information in the AR be in the "reviewed controls" or part of the "assessment log"? Maybe I'm way off on how this should be done.

Thanks!

**Comment by aj-stein-nist** [comment, marked as answer]
That's a great question. I guess it really depends on what specifically you are documenting that "tested for and confirmed compliance to a specific control." Do you mean  what you had done at the level of description in 800-53 assessment procedures (like [800-53A's](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53Ar5-draft.pdf) objectives and interview criteria for "[determining] the frequency of baseline configuration review and update is defined by [interviewing] organizational personnel with configuration management responsibilities" or the details supporting that objective and methods?

I am not sure it will be super helpful, but the [FedRAMP Guide to OSCAL-based SARs](https://raw.githubusercontent.com/GSA/fedramp-automation/090e669e2bf177e03117e50e0389d56c422d0271/documents/Guide_to_OSCAL-based_FedRAMP_Security_Assessment_Reports_(SAR).pdf) show how some of that information would look in existing FedRAMP documents (that might not be your exact use case), but the detail around the specific topic you asked about are scant.

Some more detailed examples would help better answer the question (at least for me).

**Reply by flickerfly** [reply]
Thanks for giving me some idea on what details to be looking for. I'm thinking about this in terms of automation, specifically detailed automation coming out of pipeline perhaps Terraform & Ansible code to create cloud Infrastructure and platforms, but more immediately I'm interested in creating software pipelines that outputs details to support an auditor coming behind and affirming that I'm meeting baseline configuration. Therefore, I think that's the "supporting that objective and method" you mention above. In the immediate, I expect to be given an Assessment Procedure and have to hand back how I dealt with it most recently so need to record that. If you're familiar with eMASS, short term I'm filling out a "Test Results" spreadsheet. Long term, I'm hoping to hand off OSCAL documents to eMASS (maybe via API?) so want to have this info properly placed, but haven't been able to determine that is possible yet.

Thank you also for the FedRAMP doc. I think I looked over that one, but will give it a closer look. We have been appreciating FedRAMP's example as we're trying to understand OSCAL and marry the process of software factories into security documentation. We've been regularly overwhelmed by the OSCAL details and then thankful for the thorough attention it brings to organizing the data in a re-usable and predictable manner for automation, even if we do at this point have to drop into spreadsheets at the end.

**Reply by aj-stein-nist** [reply]
> I'm thinking about this in terms of automation, specifically detailed automation coming out of pipeline perhaps Terraform & Ansible code to create cloud Infrastructure and platforms

From this, it sounds like you might be interested in the [`observation`](https://pages.nist.gov/OSCAL/reference/latest/assessment-results/xml-reference/#/assessment-results/result/observation) and its [`relevant-evidence`](https://pages.nist.gov/OSCAL/reference/latest/assessment-results/xml-reference/#/assessment-results/result/observation/relevant-evidence)

> but more immediately I'm interested in creating software pipelines that outputs details to support an auditor coming behind and affirming that I'm meeting baseline configuration. Therefore, I think that's the "supporting that objective and method" you mention above.

To me that sounds correct but is complicated. To me, this is why the SAP and SAR models deeply link and give a lot of flexibility without hard and fast rules.

Connecting the former (automation tasks I outlined from the previous part of the quote of your message) to the auditor objectives and pointing to that in the review controls log and assessment plan's activities and their tasks (that are usually much more high level and not so granular) is a complex modeling process and not so simple. Most people do not seem to show how they conceptually link the two details to the macro-level assessment goals with detailed examples in public (I have not found any so easily), so I would love to talk more with you about it!

> In the immediate, I expect to be given an Assessment Procedure and have to hand back how I dealt with it most recently so need to record that. If you're familiar with eMASS, short term I'm filling out a "Test Results" spreadsheet. Long term, I'm hoping to hand off OSCAL documents to eMASS (maybe via API?) so want to have this info properly placed, but haven't been able to determine that is possible yet.

Not 100% familiar with the eMASS Test Results, but these are the auditor-level  high-level objectives and methods from 800-53A, correct? I didn't think it surveys individual details from sampled baselines with verification procedures from one STIG check in such a baseline, no? Am I off-base?

> Thank you also for the FedRAMP doc. I think I looked over that one, but will give it a closer look. We have been appreciating FedRAMP's example as we're trying to understand OSCAL and marry the process of software factories into security documentation. We've been regularly overwhelmed by the OSCAL details and then thankful for the thorough attention it brings to organizing the data in a re-usable and predictable manner for automation, even if we do at this point have to drop into spreadsheets at the end.

There should be a lot of overlap because of [the DISA and FedRAMP reciprocity](https://disa.mil/newsandevents/2019/cloud-authorizations) so if your target is eMASS I would expect to see more an more alignment on how the data would be similarly expressed in OSCAL.

I hope we can continue the conversation, definitely feel free to come with more particular examples and we can work on them! :-)

**Reply by iMichaela** [reply]
> I'm interested in creating software pipelines that output details to support an auditor coming behind and affirming that I'm meeting baseline configuration. Therefore, I think that's the "supporting that objective and method" you mention above. In the immediate, I expect to be given an Assessment Procedure and have to hand back how I dealt with it most recently so need to record that.

@flickerfly - Do you assume an Assessment Plan (AP) exists and this plan is in OSCAL? Do you also assume that the AP is importing the SSP which, in turn, documents (imports) the baseline and describes how each component of the system satisfies the controls or, if applicable, which controls are tailored out? 

Do you also assume (as part of the CI/CD cloud deployment pipeline that the system's components' controls satisfaction is coming from different (a collection) of CDefs (aka Component Definitions)?

For example, it might be necessary to deploy several VMs as part of your system, each running the same OS but having to configure them differently. So a CDef in OSCAL might be created as a playbook for the OS. Once the VMs become part of the system, the particular configurations (tweaks) are applied and for each, the controls might be tailored as they become part of the system. The SSP would be the 'root' document that will capture the information (the tailoring), and which will also list them in the inventory. With the SSP (or particular information in the SSP) now one can generate the AP showing how the statements of each control for each VM should be assessed, then the information "how to assess" should be used by the GRC tool(s), that should be able to parse the AP to determine how to test the deployed infrastructure (or components of the infrastructure), statically and dynamically, on a staging environment. Any deficiencies need to be conveyed back to be addressed before deploying to the operational environment. OSCAL POA&M will allow you to convey the deficiencies as they are captured in the Assessment Results to the SSP/system owner and to the CDef, when applicable.

**Reply by flickerfly** [reply]
I appreciate you both pushing my knowledge along and helping me better understand what's expected.

@aj-stein-nist 

> @aj-stein-nist: Not 100% familiar with the eMASS Test Results, but these are the auditor-level high-level objectives and methods from 800-53A, correct? I didn't think it surveys individual details from sampled baselines with verification procedures from one STIG check in such a baseline, no? Am I off-base?

Yes, this sounds accurate. Part of my problem is that I'm mapping a continuous ATO concept into this also so some of the typical flow is different as a result of automation and such.

@iMichaela, I think you rightly point out the lack of following the expected flow. I'm trying to first hit the requirement of providing the eMASS Test Results document and then figure out how to model that in OSCAL as the intermediate document. While I get the dependency model on a high level, I was hoping to back into that as we build requirements demanding that data. My DevOps engineer brain is thinking strangler pattern for OSCAL adoption because we want to support machine readable and standard format of communication long-term, but our requirements are spreadsheet outputs short-term. Essentially, I'm trying to stub up as little structure as possible to hit the requirement and iterate. That's consistent with our agile development process and pushes to the MVP, then building out the rest in future iterations.

To get more into the weeds, we're trying to push solutions forward a model for a continuous ATO. Therefore, we're not evaluating the application in a running state, but during build time. We're starting off with a generic containerized application designed to run in Kubernetes. It would have a collection of various code repositories which we're working to map to CDefs. So we have a few CDefs that are evaluated through a Continuous Integration pipeline. This pipeline (sounds like that should be getting something from an AP) then returns the data to fill out the AR which is what I've been expecting. I presume the AR is loaded with the information to model out the POA&M and things like eMASS Test Results.

I'm pretty much exclusively interested in automated testing and re-affirmation of control/ap compliance relevant to applications at this point. We're assuming an inheritance of controls related to the underlaying application dependencies for now. Maybe those would be modeled as CDefs, the AP declaring the test for them being affirmation of a pre-authorized foundation and then the AR confirming that the expected foundation exists.

Another objective our team is pushing is trying to require as little application developer knowledge of OSCAL as possible (keeping it in the pipeline context) to accelerate initial adoption, but flexibility to adopt it more as their interest builds. We've built out an API for the pipeline to use based on the EasyDynamics API proposal and are working on tools to put in the pipeline to address that. Basically, trying to do the "Shift Left" thing for security process as much as possible. Therefore, we're trying to automate the creation of everything from the profile, CDefs SSP, AP, AR and such at a basic level and give them ways to send signals to that creation as part of their code repositories. A lot of things around ideas like profile selection can be presumed by context of the application.

**Reply by iMichaela** [reply]
> I think you rightly point out the lack of following the expected flow. I'm trying to first hit the requirement of providing the eMASS Test Results document and then figure out how to model that in OSCAL as the intermediate document.

AJ and I could tell you (because we've seen it happen) that if you do not have a plan for passing information from the AP to eMASS of WHAT (link to the uuids) and HOW to assess, you might have a very hard time to relate the results to WHAT and to demonstrate that the HOW defined in the AP was met.

**Reply by iMichaela** [reply]
> So we have a few CDefs that are evaluated through a Continuous Integration pipeline.

You can think of those CDefs being PROVISIONALLY A&A. Only a system owner can review the P-A&A CDefs and include them into the system. A review coming from the sys owner or authorizing official is a must because only he/she can assess the risk. Any tailoring will require a  Re-A&A. An assessment of inherited controls and stacked solutions is a must as well, especially for CDEfs that rely on the environment for defense.

**Reply by iMichaela** [reply]
> Another objective our team is pushing is trying to require as little application developer knowledge of OSCAL as possible (keeping it in the pipeline context) to accelerate initial adoption,

In my experience, this can be achieved with simple wish lists or questions or checklists mapped on the back-end to OSCAL-formatted 'answeres'/code.

**Reply by flickerfly** [reply]
@iMichaela Thank you for all those thoughts!

**Reply by aj-stein-nist** [reply]
Also, personally I'd love your attention on usnistgov/OSCAL#1058 as it evolves, as this further examination of this topic. Not necessarily your immediate question, but more to come on how to make this more specific (it's already possible in OSCAL with SAP and SAR) before SAP and SAR time.
